core/src/main/scala/kafka/log/Log.scala

```diff
@@ -380,32 +380,31 @@ class Log(@volatile var dir: File,
     * starts from a snapshot that is taken strictly before the log end
     * offset. Consequently, we need to process the tail of the log to update
     * the mapping.
-    *
-    * @param lastOffset
-    *
-    * @return An instance of ProducerIdMapping
     */
   private def buildAndRecoverPidMap(lastOffset: Long) {
     lock synchronized {
+      info(s"Recovering PID mapping from offset $lastOffset for partition $topicPartition")
       val currentTimeMs = time.milliseconds
       pidMap.truncateAndReload(lastOffset, currentTimeMs)
       logSegments(pidMap.mapEndOffset, lastOffset).foreach { segment =>
         val startOffset = math.max(segment.baseOffset, pidMap.mapEndOffset)
         val fetchDataInfo = segment.read(startOffset, Some(lastOffset), Int.MaxValue)
-        val records = fetchDataInfo.records
-        records.batches.asScala.foreach { batch =>
-          if (batch.hasProducerId) {
-            // TODO: Currently accessing any of the batch-level headers other than the offset
-            // or magic causes us to load the full entry into memory. It would be better if we
-            // only loaded the header
-            val numRecords = (batch.lastOffset - batch.baseOffset + 1).toInt
-            val pidEntry = ProducerIdEntry(batch.producerEpoch, batch.lastSequence, batch.lastOffset,
-              numRecords, batch.maxTimestamp)
-            pidMap.load(batch.producerId, pidEntry, currentTimeMs)
+        if (fetchDataInfo != null) {
+          fetchDataInfo.records.batches.asScala.foreach { batch =>
+            if (batch.hasProducerId) {
+              // TODO: Currently accessing any of the batch-level headers other than the offset
+              // or magic causes us to load the full entry into memory. It would be better if we
+              // only loaded the header
+              val numRecords = (batch.lastOffset - batch.baseOffset + 1).toInt
```

Comment:
Why is it safe to assume that no compaction has happened? Probably worth a comment.